ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,437 @@
+    char **arg_strings = opal_argv_split(ompi_mpi_spc_attach_string, ',');
+    int num_args       = opal_argv_count(arg_strings);
+
+    /* If there is only one argument and it is 'all', then all counters
+     * should be turned on.  If the size is 0, then no counters will be enabled.
+     */
+    if(1 == num_args) {
+        if(strcmp(arg_strings[0], "all") == 0) {
+            all_on = 1;
+        }
+    }
+
+    /* Turn on only the counters that were specified in the MCA parameter */
+    for(i = 0; i < OMPI_SPC_NUM_COUNTERS; i++) {
+        if(all_on) {
+            SET_SPC_BIT(ompi_spc_attached_event, i);
+            mpi_t_enabled = true;
+            found++;
+        } else {
+            /* Note: If no arguments were given, this will be skipped */
+            for(j = 0; j < num_args; j++) {
+                if( 0 == strcmp(ompi_spc_events_names[i].counter_name, arg_strings[j]) ) {
+                    SET_SPC_BIT(ompi_spc_attached_event, i);
+                    mpi_t_enabled = true;
+                    found++;
+                    break;
+                }
+            }
+        }
+
+        /* ########################################################################
+         * ################## Add Timer-Based Counter Enums Here ##################
+         * ########################################################################
+         */
+        CLEAR_SPC_BIT(ompi_spc_timer_event, i);
+
+        /* Registers the current counter as an MPI_T pvar regardless of whether it's been turned on or not */
+        ret = mca_base_pvar_register("ompi", "runtime", "spc", ompi_spc_events_names[i].counter_name, ompi_spc_events_names[i].counter_description,
+                                     OPAL_INFO_LVL_4, MPI_T_PVAR_CLASS_SIZE,
+                                     MCA_BASE_VAR_TYPE_UNSIGNED_LONG_LONG, NULL, MPI_T_BIND_NO_OBJECT,
+                                     MCA_BASE_PVAR_FLAG_READONLY | MCA_BASE_PVAR_FLAG_CONTINUOUS,
+                                     ompi_spc_get_count, NULL, ompi_spc_notify, NULL);
+
+        /* Check to make sure that ret is a valid index and not an error code.
+         */
+        if( ret >= 0 ) {
+            if( mpi_t_offset == -1 ) {
+                mpi_t_offset = ret;
+            }
+        }
+        if( (ret < 0) || (ret != (mpi_t_offset + found - 1)) ) {
+            mpi_t_enabled = false;
+            opal_show_help("help-mpi-runtime.txt", "spc: MPI_T disabled", true);
+            break;
+        }
+    }
+    /* If this is a timer event, sent the corresponding timer_event entry to 1 */
+    SET_SPC_BIT(ompi_spc_timer_event, OMPI_MATCH_TIME);
+    opal_argv_free(arg_strings);
+}
+
+/* Gathers all of the SPC data onto rank 0 of MPI_COMM_WORLD and prints out all
+ * of the counter values to stdout.
+ */
+void ompi_spc_dump(void)
+{
+    int i, j, world_size, offset;
+    long long *recv_buffer = NULL, *send_buffer;
+
+    ompi_communicator_t *comm = &ompi_mpi_comm_world.comm;
```

Comment:
Per my prior comment:

> It may not be safe to use MPI_COMM_WORLD here. I think you might want to dup COMM_WORLD during your startup and then use that here in the dump (and then be sure to free that communicator when you're done with it).

Reply:
I'll fix this this evening.